flow/connectors/clickhouse/normalize.go

```diff
@@ -524,91 +525,93 @@ func (c *ClickHouseConnector) NormalizeRecords(
-			tbl,
-			req.TableNameSchemaMapping,
-			req.TableMappings,
-			endBatchID,
-			lastNormBatchIDForTable,
-			enablePrimaryUpdate,
-			sourceSchemaAsDestinationColumn,
-			req.Env,
-			rawTbl,
-			c.chVersion,
-			c.Config.Cluster != "",
-			req.SoftDeleteColName,
-			req.Version,
-		)
-		query, err := queryGenerator.BuildQuery(ctx)
-		if err != nil {
-			close(queries)
-			c.logger.Error("[clickhouse] error while building insert into select query",
-				slog.String("table", tbl),
-				slog.Int64("endBatchID", endBatchID),
-				slog.Int64("lastNormBatchID", lastNormBatchIDForTable),
-				slog.Any("error", err))
-			return model.NormalizeResponse{}, fmt.Errorf("error while building insert into select query for table %s: %w", tbl, err)
-		}
+			// Skip batches already normalized for this table. This can happen if a previous normalization run partially succeeded.
+			lastNormBatchIDForTable = max(lastNormBatchID, lastNormBatchIDForTable)
+			if lastNormBatchIDForTable >= endBatchID {
+				c.logger.Info("[clickhouse] latest batch already synced to destination, skipping",
+					"table", tbl, "lastNormBatchID", lastNormBatchIDForTable, "endBatchID", endBatchID)
+				continue
+			}
 
-		select {
-		case queries <- queryInfo{
-			table:           tbl,
-			query:           query,
-			lastNormBatchID: lastNormBatchIDForTable,
-		}:
-		case <-errCtx.Done():
-			close(queries)
-			c.logger.Error("[clickhouse] context canceled while inserting data to ClickHouse",
-				slog.Any("error", errCtx.Err()),
-				slog.Any("cause", context.Cause(errCtx)))
-			return model.NormalizeResponse{}, context.Cause(errCtx)
+			queryGenerator := NewNormalizeQueryGenerator(
+				tbl,
+				req.TableNameSchemaMapping,
+				req.TableMappings,
+				endBatchID,
+				lastNormBatchIDForTable,
+				enablePrimaryUpdate,
+				sourceSchemaAsDestinationColumn,
+				req.Env,
+				rawTbl,
+				c.chVersion,
+				c.Config.Cluster != "",
+				req.SoftDeleteColName,
+				req.Version,
+			)
+			query, err := queryGenerator.BuildQuery(ctx)
+			if err != nil {
+				c.logger.Error("[clickhouse] error while building insert into select query",
+					slog.String("table", tbl),
+					slog.Int64("endBatchID", endBatchID),
+					slog.Int64("lastNormBatchID", lastNormBatchIDForTable),
+					slog.Any("error", err))
+				return model.NormalizeResponse{}, fmt.Errorf("error while building insert into select query for table %s: %w", tbl, err)
+			}
+
+			select {
+			case queriesCh <- queryInfo{
+				table:           tbl,
+				query:           query,
+				lastNormBatchID: lastNormBatchIDForTable,
+			}:
+			case <-errCtx.Done():
+				c.logger.Error("[clickhouse] context canceled while inserting data to ClickHouse",
+					slog.Any("error", errCtx.Err()),
+					slog.Any("cause", context.Cause(errCtx)))
+				return model.NormalizeResponse{}, context.Cause(errCtx)
+			}
 		}
-	}
-	close(queries)
+		return model.NormalizeResponse{
+			StartBatchID: lastNormBatchID + 1,
+			EndBatchID:   endBatchID,
```

Comment:
lastNormBatchID and endBatchID are not modified in the helper function so it doesn't need to be involved with the response return?